targets/TARGET_STM/stm_spi_api.c

```diff
@@ -777,75 +800,225 @@ static inline int datasize_to_transfer_bitshift(uint32_t DataSize)
-     *  but this will increase performances significantly
-     */
-
 #if TARGET_STM32H7
-    /* Master transfer start */
-    LL_SPI_StartMasterTransfer(SPI_INST(obj));
+    status = ((__HAL_SPI_GET_FLAG(handle, SPI_FLAG_TXC) == RESET) ? 1 : 0);
+#else /* TARGET_STM32H7 */
+    status = ((__HAL_SPI_GET_FLAG(handle, SPI_FLAG_BSY) != RESET) ? 1 : 0);
+#endif /* TARGET_STM32H7 */
+    return status;
+}
 
+/**
+ * Wait till SPI master interface is writable.
+ */
+static inline void msp_wait_writable(spi_t *obj)
+{
+#if TARGET_STM32H7
     /* Wait TXP flag to transmit data */
     while (!LL_SPI_IsActiveFlag_TXP(SPI_INST(obj)));
-#else
+#else /* TARGET_STM32H7 */
     /* Wait TXE flag to transmit data */
     while (!LL_SPI_IsActiveFlag_TXE(SPI_INST(obj)));
+#endif /* TARGET_STM32H7 */
+}
 
+/**
+ * Wait till SPI master interface is readable.
+ */
+static inline void msp_wait_readable(spi_t *obj)
+{
+#if TARGET_STM32H7
+    /* Wait for RXP flag */
+    while (!LL_SPI_IsActiveFlag_RXP(SPI_INST(obj)));
+#else /* TARGET_STM32H7 */
+    /* Wait for RXNE flag before reading */
+    while (!LL_SPI_IsActiveFlag_RXNE(SPI_INST(obj)));
 #endif /* TARGET_STM32H7 */
+}
 
-    /* Transmit data */
+/**
+ * Write data to SPI register.
+ */
+static inline void msp_write_data(spi_t *obj, int value, int bitshift)
+{
     if (bitshift == 1) {
         LL_SPI_TransmitData16(SPI_INST(obj), (uint16_t)value);
 #ifdef HAS_32BIT_SPI_TRANSFERS
     } else if (bitshift == 2) {
         LL_SPI_TransmitData32(SPI_INST(obj), (uint32_t)value);
-#endif
+#endif /* HAS_32BIT_SPI_TRANSFERS */
     } else {
         LL_SPI_TransmitData8(SPI_INST(obj), (uint8_t)value);
     }
+}
 
-#if TARGET_STM32H7
-    /* Wait for RXP or end of Transfer */
-    while (!LL_SPI_IsActiveFlag_RXP(SPI_INST(obj)));
-#else /* TARGET_STM32H7 */
-    /* Wait for RXNE flag before reading */
-    while (!LL_SPI_IsActiveFlag_RXNE(SPI_INST(obj)));
-#endif /* TARGET_STM32H7 */
-
-    /* Read received data */
+/**
+ * Read data from SPI register.
+ */
+static inline int msp_read_data(spi_t *obj, int bitshift)
+{
     if (bitshift == 1) {
         return LL_SPI_ReceiveData16(SPI_INST(obj));
 #ifdef HAS_32BIT_SPI_TRANSFERS
     } else if (bitshift == 2) {
         return LL_SPI_ReceiveData32(SPI_INST(obj));
-#endif
+#endif /* HAS_32BIT_SPI_TRANSFERS */
     } else {
         return LL_SPI_ReceiveData8(SPI_INST(obj));
     }
 }
 
+
+/**
+ * Transmit and receive SPI data in bidirectional mode.
+ *
+ * @param obj spi object
+ * @param tx_buffer byte-array of data to write to the device
+ * @param tx_length number of bytes to write, may be zero
+ * @param rx_buffer byte-array of data to read from the device
+ * @param rx_length number of bytes to read, may be zero
+ * @return number of transmitted and received bytes or negative code in case of error.
+ */
+static int spi_master_one_wire_write(spi_t *obj, const char *tx_buffer, int tx_length, char *rx_buffer, int rx_length)
```

Comment:
maybe this should actually be called "one_wire_transfer" ? as it does manage tx and rx

Reply:
Done. Yes, "_transfer" suffix is more meaningful.
note: I used "_write" suffix for consistency with mbed function `spi_master_block_write` function (it has `_write` suffix and manages rx and tx).